sympy/geometry/ellipse.py

```diff
@@ -1339,28 +1341,79 @@ class Circle(Ellipse):
     >>> c2.hradius, c2.vradius, c2.radius, c2.center
     (sqrt(2)/2, sqrt(2)/2, sqrt(2)/2, Point2D(1/2, 1/2))
 
+    >>> # a circle object is returned
+    >>> from sympy.abc import x, y, a, b
+    >>> Circle(x ** 2 + y ** 2 - 25)
+    Circle(Point2D(0, 0), 5)
+    >>> Circle(a ** 2 + b ** 2 - 25, x='a', y='b')
+    Circle(Point2D(0, 0), 5)
+
     """
 
     def __new__(cls, *args, **kwargs):
-        c, r = None, None
-        if len(args) == 3:
-            args = [Point(a, dim=2) for a in args]
-            if Point.is_collinear(*args):
-                raise GeometryError(
-                    "Cannot construct a circle from three collinear points")
-            from .polygon import Triangle
-            t = Triangle(*args)
-            c = t.circumcenter
-            r = t.circumradius
-        elif len(args) == 2:
-            # Assume (center, radius) pair
-            c = Point(args[0], dim=2)
-            r = sympify(args[1])
-
-        if not (c is None or r is None):
-            return GeometryEntity.__new__(cls, c, r, **kwargs)
-
-        raise GeometryError("Circle.__new__ received unknown arguments")
+
+        if len(args) == 1 and isinstance(args[0], Expr):
+            x = kwargs.get('x', 'x')
+            y = kwargs.get('y', 'y')
+            equation = args[0]
+
+            def find(x_, equation_):
+                free = equation_.free_symbols
+                xs = [i for i in free if (i.name if type(x_) is str else i) == x_]
+                if not xs:
+                    raise ValueError('could not find %s' % x_)
+                if len(xs) != 1:
+                    raise ValueError('ambiguous %s' % x_)
+                return xs[0]
+
+            x = find(x, equation)
+            y = find(y, equation)
+
+            try:
+                center_x = solve(diff(equation, x))[0]
```

Comment:
What if the user passes a cubic?

Consider using something form polys. I think there is a biquadratic solver.

Reply:
@smichr 
Passing cubics will result in `GeometricError`. Here's how it shows up : 
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> from sympy.abc import *
>>> Circle(x ** 3 + y ** 3 - 25)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/avi/sympy/sympy/geometry/ellipse.py", line 1388, in __new__
    raise GeometryError("The given equation is not that of a circle.")
sympy.geometry.exceptions.GeometryError: The given equation is not that of a circle.
>>> Circle(x ** 2 + y ** 3 - 25)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/avi/sympy/sympy/geometry/ellipse.py", line 1388, in __new__
    raise GeometryError("The given equation is not that of a circle.")
sympy.geometry.exceptions.GeometryError: The given equation is not that of a circle.
```
So passing cubics isn't a problem.

Reply:
missed that part, sorry.

Reply:
BTW, linear_coeffs could be used here, too:
```
x2 = Dummy('%s**2' % x)
y2 = Dummy('%s**2' % y)
eq = eq.subs(x**2, x2).subs(y**2, y2)
a, b, c, d, e = linear_coeff(_mexpand(eq, recursive=True), x2, y2, x, y)
```